### ISSUE TITLE
Performance degradation due to license checks

### ISSUE THREAD

**Issue by bhalbright** [DuendeSoftware/Support#1161]
**Which version of Duende IdentityServer are you using?**
6.2.3

**Which version of .NET are you using?**
7

**Describe the bug**
We notice performance degradation of Duende Identity Server at scale. From using a profiler, the most time is spent in the Duende.IdentityServer.Validation.LicenseValidator.ValidateClient(string) method, where the majority of that time is in ConcurrentDictionary operations. The end result is that response times to the token endpoint that normally would be sub-second or less start to take 10-20 seconds.

**To Reproduce**
1000s of concurrent requests

Steps to reproduce the behavior.

**Expected behavior**
License checks to impact performance in a negligible manner, even at scale.

**Log output/exception with stacktrace**

```
no log output of note
```

**Additional context**

We will move to version 7 and .net 8 soon, but I noticed the license check code is basically the same, though it was moved around some, so I think the issue would still be present in the latest version.

Here is a screenshot of the profiling done, this was 

![image](https://github.com/DuendeSoftware/Support/assets/32200834/1272f328-0a6a-4e91-9bf1-6fb80f56de31)

**Comment by AndersAbel** [comment]
Thanks for reporting this. I inspected the locking behaviour in `ConcurrentDictionary` and it looks like the locking behaviour was optimized for reading/writing specific keys while the `Count` property is not that efficient.

The code was updated in version 7. If you have a license with unlimited number of clients the problem should no longer appear. But it could still cause issues on the starter and business editions. I'm transferring this issue to the IdentityServer repository as a work item.

**Comment by josephdecock** [comment]
Also, we need to decide if changes would be backported to the 6.3 release or if we would only apply these to the 7.0 branch. That would basically come down to 
- We need to solve your problem @bhalbright, so let us know if including this in your planned upgrade to 7.0 works for you
- We need to understand how this is likely to affect other 6.x users (hence my questions above)

**Comment by brockallen** [comment]
Hi @bhalbright. If we were able to put a 7.0.x patch preview out, would you be able to do your load test against it?

**Comment by bhalbright** [comment]
> @AndersAbel, I don't doubt your analysis of the concurrent dictionary implementation, but I also wonder if there is something interesting or unusual about the environment or usage pattern here. This is the first time we've heard about this issue, so why haven't other users experienced this? @bhalbright can you characterize what you're seeing in more detail? My theory is that there is a threshold of concurrency where the performance problems transition from being not really noticeable to a cascading failure. Is that consistent with your observations of your system?

I agree with this assessment, we have a customer with 100s of thousands of clients there seems to be some threshold of concurrency where the performance degrades, though I don't know the exact tipping point. We added a second instance of the service and then the response times got into the normal range (sub-second). However, I was able to reproduce the slow down in a K6 suite we have which is where I got the screenshot of the profiler.

**Comment by bhalbright** [comment]
> Also, we need to decide if changes would be backported to the 6.3 release or if we would only apply these to the 7.0 branch. That would basically come down to
> 
> * We need to solve your problem @bhalbright, so let us know if including this in your planned upgrade to 7.0 works for you
> * We need to understand how this is likely to affect other 6.x users (hence my questions above)

We would be fine with any fix going to 7, we are working on the migration to .net 8 right now and this would be included in that work.

**Comment by bhalbright** [comment]
> Hi @bhalbright. If we were able to put a 7.0.x patch preview out, would you be able to do your load test against it?

Yes, I should be able to.

**Comment by brockallen** [comment]
Ok great! We've just pushed a 7.0.2-preview.1 to NuGet. Let us know how it goes. Thanks @bhalbright.

**Comment by bhalbright** [comment]
> Ok great! We've just pushed a 7.0.2-preview.1 to NuGet. Let us know how it goes. Thanks @bhalbright.

Thanks, I have been on vacation a bit. I hope to be able to try this out this week.

**Comment by bhalbright** [comment]
@brockallen 

I ran our load test against 7.0.2-preview.1 and 7.0.1 for comparison since my prior test was against 6.2.3. The change was very successful. Thanks again! Offhand how far off would you think this change making into an official release would be?


**7.0.2-preview.1**
little time in ValidateClient, I noticed the response times to the token endpoint remained very fast, even with ~30k requests per minute

![image](https://github.com/DuendeSoftware/Support/assets/32200834/ee7f821a-405a-4122-9295-e3063863e2a4)


7.0.1
same/similar result to 6.2.3, most of the time is spent in the ValidateClient/ConcurrentDictionary and we see response times degrade.

![image](https://github.com/DuendeSoftware/Support/assets/32200834/a86b5687-2e49-407f-8573-aa3cb6dd7933)

**Comment by brockallen** [comment]
>  Offhand how far off would you think this change making into an official release would be?

If you are asking for a 7.0.2 patch, we can get that out today. Let us know, please. Thanks!

**Comment by bhalbright** [comment]
> > Offhand how far off would you think this change making into an official release would be?
> 
> If you are asking for a 7.0.2 patch, we can get that out today. Let us know, please. Thanks!

That would awesome if possible, thanks!
